lib/membership_pair_widget.dart

```diff
@@ -0,0 +1,175 @@
+    final onWebRtcOfferCreated = (WebRtcOffer webRtcOffer) {
+      membershipPairState.value = membershipPairState.value.copyWith(webRtcOffer: webRtcOffer);
+    };
+    final onWebRtcAnswerCreated = (WebRtcAnswer webRtcAnswer) {
+      membershipPairState.value = membershipPairState.value.copyWith(webRtcAnswer: webRtcAnswer);
+    };
+    final onWebRtcIceCandidateCreated = (WebRtcIceCandidate webRtcIceCandidate) {
+      final webRtcIceCandidates = [...membershipPairState.value.webRtcIceCandidates, webRtcIceCandidate];
+      membershipPairState.value = membershipPairState.value.copyWith(webRtcIceCandidates: webRtcIceCandidates);
+    };
+
+    final membershipPairChannelState = useState(
+      MembershipPairChannel(
+        cable: cable,
+        id: membershipPairEntry.membershipPairId,
+        onConnected: updateMembershipPair,
+        onWebRtcOfferCreated: onWebRtcOfferCreated,
+        onWebRtcAnswerCreated: onWebRtcAnswerCreated,
+        onWebRtcIceCandidateCreated: onWebRtcIceCandidateCreated
+      )
+    );
+
+    useEffect(() {
+      membershipPairChannelState.value.subscribe();
+      return () => membershipPairChannelState.value.unsubscribe();
+    });
+
+    final localStreamFuture = useProvider(localStreamProvider.future);
+    final peerConnectionState = useState<RTCPeerConnection>(null);
+    final peerConnectionSignalingStateState = useState<RTCSignalingState>(null);
+    final remoteStreamState = useState<MediaStream>(null);
+
+    final onSignalingState = (RTCSignalingState signalingState) => peerConnectionSignalingStateState.value = signalingState;
+    final onIceCandidate = (RTCIceCandidate iceCandidate) {
+      membershipPairChannelState.value.createWebRtcIceCandidate(
+        sdp: iceCandidate.candidate,
+        sdpMid: iceCandidate.sdpMid,
+        sdpMlineIndex: iceCandidate.sdpMlineIndex
+      );
+    };
+    final onAddStream = (MediaStream stream) => remoteStreamState.value = stream;
+
+    useEffect(() {
+        Future.wait([
+          localStreamFuture,
+          createPeerConnection(PEER_CONFIG, {})
+        ]).then((resolvedFutures) {
+          final MediaStream localStream = resolvedFutures[0];
+          final RTCPeerConnection peerConnection = resolvedFutures[1];
+
+          peerConnection.addStream(localStream);
+          peerConnection.onSignalingState = onSignalingState;
+          peerConnection.onIceCandidate = onIceCandidate;
+          peerConnection.onAddStream = onAddStream;
+
+          peerConnectionState.value = peerConnection;
+        });
+
+        return () => peerConnectionState.value?.dispose();
+    }, []);
+
+    useEffect(() {
+      if (peerConnectionState.value == null) {
+        return;
+      }
+
+      () async {
+        final peerConnection = peerConnectionState.value;
+        if (membershipPairEntry.older && peerConnectionSignalingStateState.value == RTCSignalingState.RTCSignalingStateStable) {
```

Comment:
this feels like a big antipattern, like I'm modeling procedural behavior in a series of if-thens where we memoize how far we've made it through this workflow and come back to pick up where we've left off.

but I couldn't think of a better pattern.

Reply:
Yea so I think if lift most of these use effect hooks up into a state notifier view model-y thing then we can right this all prcedure-y and feel better about it.